File: src/aiobserve/extract/pricing.py
"""What an API call cost, from a price table we maintain.

This table is **ours**, not Claude Code's. A model missing from it is a gap in our list, not
a schema change to surface, so `compute_cost` returns `None` and the caller records the
model name with no cost — a queryable absence that can be filled in later. Crashing here
would kill a backfill the day a new model ships.

Prices are USD per million tokens, read from
<https://platform.claude.com/docs/en/about-claude/pricing> on **2026-08-07**. Nothing in the
test suite can check them against that page; re-read it and update the date when you touch
the table.

Two published modifiers are deliberately absent, because no recorded call uses them: fast
mode and US-only inference. Across the mycelia corpus's ~290,000 assistant records (scanned
2026-08-07), `usage.speed` is `"standard"` on half and absent on the rest, never `"fast"`,
and `usage.inference_geo` is `"not_available"` wherever it appears, never `"us"`. A call
carrying either would price low, so add the multiplier when one shows up.
"""

from typing import NamedTuple

# Claude Code's own placeholder replies — an interrupt notice, a cancelled request — are
# recorded as assistant records under this model name. They report zero tokens.
SYNTHETIC_MODEL = "<synthetic>"

# Cache multipliers on the base input rate, from the pricing page's caching section. The
# TTL is what separates them: a 1-hour cache pays off after two reads, a 5-minute one after
# one.
_CACHE_READ = 0.1
_CACHE_WRITE_5M = 1.25
_CACHE_WRITE_1H = 2.0

_PER_MILLION = 1_000_000


class ModelPrice(NamedTuple):
    """One model's base rates, USD per million tokens. Cache rates derive from `input`."""

    input: float
    output: float


class TokenUsage(NamedTuple):
    """The token counts one reply reported, as the pricing table charges them."""

    input: int
    output: int
    cache_read: int
    cache_creation: int
    # The cache-creation total split by TTL. Both None when the reply reported no split at
    # all, which prices the whole write at the 5-minute rate.
    cache_5m: int | None
    cache_1h: int | None


# Every model the mycelia corpus records, plus the placeholder. Keyed by the exact
# `message.model` string, since that is what the transcript carries.
PRICES: dict[str, ModelPrice] = {
    SYNTHETIC_MODEL: ModelPrice(input=0.0, output=0.0),
    "claude-fable-5": ModelPrice(input=10.0, output=50.0),
    "claude-opus-5": ModelPrice(input=5.0, output=25.0),
    "claude-opus-4-8": ModelPrice(input=5.0, output=25.0),
    "claude-opus-4-1-20250805": ModelPrice(input=15.0, output=75.0),
    # Introductory pricing, in effect through 2026-08-31; it rises to $3/$15 on
    # 2026-09-01. The table is flat, so a call recorded after that date will price low
    # until this line is split by effective date.
    "claude-sonnet-5": ModelPrice(input=2.0, output=10.0),
    "claude-sonnet-4-6": ModelPrice(input=3.0, output=15.0),
    "claude-haiku-4-5-20251001": ModelPrice(input=1.0, output=5.0),
}


def compute_cost(model: str, tokens: TokenUsage) -> float | None:
    """What one reply cost in USD, or None when the table does not price its model."""
    price = PRICES.get(model)
    if price is None:
        return None
    if tokens.cache_5m is None or tokens.cache_1h is None:
        write = tokens.cache_creation * _CACHE_WRITE_5M
    else:
        write = tokens.cache_5m * _CACHE_WRITE_5M + tokens.cache_1h * _CACHE_WRITE_1H
    return (
        tokens.input * price.input
        + tokens.output * price.output
        + tokens.cache_read * price.input * _CACHE_READ
        + write * price.input
    ) / _PER_MILLION

File: rust/crates/hyphae-extract/tests/pricing.rs
//! What a recorded reply cost, from our own price table.
//!
//! The port of `tests/extract/test_pricing.py`. The table is ours, not Claude Code's — a
//! model it lacks is a gap in our list, not a schema change, so the leaves below pin
//! behaviour on both sides of that line. What they cannot pin is whether the numbers match
//! Anthropic's published prices: no seam reaches that, and a test asserting a constant against
//! itself proves nothing. `pricing.py` records the check date. The claim that the price table
//! and the window table answer one census is a unit test beside them, the price table being
//! private.

use hyphae_extract::pricing::{SYNTHETIC_MODEL, TokenUsage, compute_cost, split_cost};

/// Lifted verbatim from `tests/fixtures/spine/`, CC 2.1.221 — the usage of
/// `msg_011CdmMjFXDofyYSMxYtXa5n`, a `claude-fable-5` reply that put its whole cache write on
/// the 1-hour TTL.
const SPINE_SPLIT: TokenUsage = TokenUsage {
    input: 2,
    output: 415,
    cache_read: 9_768,
    cache_creation: 20_257,
    cache_5m: Some(0),
    cache_1h: Some(20_257),
};

/// Two dollar amounts agree to the cent a thousand times over, which is as close as two
/// orderings of the same float addition get.
#[track_caller]
fn close(charged: f64, expected: f64) {
    assert!(
        (charged - expected).abs() < 1e-9,
        "{charged} is not {expected}"
    );
}

/// Input, output, cache read and cache write each price at their own rate.
///
/// The four are also readable one at a time: the viewer's popover prints a legend saying where
/// a phase's dollars went, and a legend derived from anything but this arithmetic would be a
/// second answer to what a call cost (`docs/viewer.md`).
#[test]
fn a_reply_is_priced_by_its_model_and_its_four_token_kinds() {
    // If a Fable 5 reply reports the four token kinds — $10/MTok in, $50/MTok out, cache
    // reads at 0.1x input and a 1-hour cache write at 2x input...
    let cost = compute_cost("claude-fable-5", &SPINE_SPLIT).expect("the table prices fable");
    let split = split_cost("claude-fable-5", &SPINE_SPLIT).expect("the table prices fable");

    // ...then each is charged at its own rate and the four are summed.
    let kinds = [
        2.0 * 10.0,            // input
        415.0 * 50.0,          // output
        9_768.0 * 10.0 * 0.1,  // cache read
        20_257.0 * 10.0 * 2.0, // 1-hour cache write
    ];
    close(cost, kinds.iter().sum::<f64>() / 1_000_000.0);
    close(cost, 0.435_678);
    // ...and the split hands back the same four separately, in the same USD the total is in.
    for (charged, kind) in [
        split.input,
        split.output,
        split.cache_read,
        split.cache_write,
    ]
    .iter()
    .zip(kinds)
    {
        close(*charged, kind / 1_000_000.0);
    }
    // The total is the split summed, so the legend can never disagree with the badge above it.
    close(
        split.input + split.output + split.cache_read + split.cache_write,
        cost,
    );
}

/// A 1-hour cache write costs 2x input; a 5-minute one costs 1.25x.
#[test]
fn the_cache_write_splits_by_ttl() {
    // If two replies write the same number of cache tokens under different TTLs...
    let written = |five, hour| TokenUsage {
        input: 0,
        output: 0,
        cache_read: 0,
        cache_creation: 1_000,
        cache_5m: Some(five),
        cache_1h: Some(hour),
    };

    // ...then the 1-hour write costs 2x the base input rate and the 5-minute one 1.25x, so
    // the same tokens cost 60% more on the longer TTL.
    close(
        compute_cost("claude-opus-5", &written(0, 1_000)).expect("the table prices opus"),
        1_000.0 * 5.0 * 2.0 / 1_000_000.0,
    );
    close(
        compute_cost("claude-opus-5", &written(1_000, 0)).expect("the table prices opus"),
        1_000.0 * 5.0 * 1.25 / 1_000_000.0,
    );
}

/// When a reply reports no TTL split, its whole cache write prices as 5-minute.
///
/// INVENTED shape: every assistant record in the mycelia corpus carries `usage.cache_creation`
/// (scanned 2026-08-07), so the unsplit reading is a fallback we chose, not one the corpus
/// shows.
#[test]
fn an_unsplit_cache_write_is_charged_at_the_five_minute_rate() {
    let unsplit = TokenUsage {
        input: 0,
        output: 0,
        cache_read: 0,
        cache_creation: 1_000,
        cache_5m: None,
        cache_1h: None,
    };

    close(
        compute_cost("claude-opus-5", &unsplit).expect("the table prices opus"),
        1_000.0 * 5.0 * 1.25 / 1_000_000.0,
    );
}

/// An unpriced model reports no cost at all, so a query can find it and fill it in.
///
/// The fail-fast rule guards Claude Code's schema, not our price list: a model released
/// mid-backfill must not kill the backfill, and pricing it at zero would report a free
/// session — the prior importer's bug.
#[test]
fn a_model_the_table_lacks_costs_nothing_rather_than_zero() {
    assert_eq!(compute_cost("claude-mythos-9", &SPINE_SPLIT), None);
    // And the split says the same nothing, so a legend cannot print four zeroes where the
    // badge beside it printed no number at all.
    assert!(split_cost("claude-mythos-9", &SPINE_SPLIT).is_none());
}

/// Claude Code's own placeholder replies are priced, and priced at nothing.
///
/// The 205 `<synthetic>` records in the corpus all report zero tokens, so the zero is
/// over-determined — but the table names the model so the cost is a stated zero rather than an
/// unpriced null.
#[test]
fn a_synthetic_reply_costs_zero() {
    assert_eq!(compute_cost(SYNTHETIC_MODEL, &SPINE_SPLIT), Some(0.0));
}
